### PR TITLE
Speed up the agent

### PR DESCRIPTION
**Pull request by jrj-d**
Sometimes, the queue of events created by Spark in front of the Delight agent would grow and hit the `spark.scheduler.listenerbus.eventqueue.capacity`.
The reason was that the agent would not ingest the events fast enough.

This would happen when many small tasks (a few ms each) would run in parallel.

The main culprit was the conversion to json and compaction that was done in a blocking way in `logEvent` (5x to 20x speedup, hard to say precisely).

This operation is now done at the very end, when the events are sent over the network.

## How to test

Use the `just` commands in the project. This won't prove the speedup but you'll be able to validate that apps still appear in Delight.

### CODE REVIEW

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 1]
```diff
@@ -287,13 +291,15 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
...
+  private def startIfNecessary(): Unit = Utils.time(shouldLogDuration, "startIfNecessary") {
     if(started.compareAndSet(false, true)) {
       val pollingThread = new Thread {
         override def run() {
           while (true) {
+            val start = System.currentTimeMillis()
             publishPendingEvents()
-            Thread.sleep(currentPollingInterval.toMillis)
+            val end = System.currentTimeMillis()
+            Thread.sleep(math.max(currentPollingInterval.toMillis - (end - start), 0))
```
Comment (ImpSy) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: Could we  mutualize the creation of a thread w/ sleep in an Utils method ?
Because we repeat almost the same code for the polling thread and the heartbeat thread

File: src/main/scala/co/datamechanics/delight/Configs.scala [hunk 2]
```diff
@@ -46,11 +46,17 @@ object Configs {
     sparkConf.getDouble("spark.delight.waitForPendingPayloadsSleepIntervalSecs", 1).seconds
   }
 
+  def logDuration(sparkConf: SparkConf): Boolean = {
+    sparkConf.getBoolean("spark.delight.logDuration", false)
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/Configs.scala:50]: nit: Would be nice to have a value for this configuration property in a Scala object.

File: src/main/scala/co/datamechanics/delight/DelightListener.scala [hunk 3]
```diff
@@ -3,31 +3,38 @@ import java.util.concurrent.atomic.AtomicBoolean
...
-import org.json4s.jackson.JsonMethods.{compact, render}
+import org.apache.spark.SparkConf
 
 
 class DelightListener(sparkConf: SparkConf) extends SparkListener with Logging {
 
+  /**
+   * Adds Delight version to the Spark config
+   */
+  sparkConf.set("spark.delight.version", Configs.delightVersion)
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightListener.scala:14]: nit: A Config object with the configuration property would be awesome.

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 4]
```diff
@@ -262,7 +263,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
         publishPayload(
           StreamingPayload(
             dmAppId,
-            firstEvents,
+            firstEvents.map(e => compact(render(JsonProtocolProxy.jsonProtocol.sparkEventToJson(e)))),
```
Comment (geoffroynicolas) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala:267]: **question:** Why is it faster if in the end we have to loop over all events to serialize them ? It is not reducing the number of serialization operations right ?

File: src/main/scala/co/datamechanics/delight/DelightListener.scala [hunk 5]
```diff
@@ -3,31 +3,38 @@ import java.util.concurrent.atomic.AtomicBoolean
...
 
-  private def logEvent(event: SparkListenerEvent, flush: Boolean = false, blocking: Boolean = false): Unit = {
+  private def logEvent(event: SparkListenerEvent, flush: Boolean = false, blocking: Boolean = false): Unit = Utils.time(
+    shouldLogDuration, "logEvent"
+  ) {
     sendLogStartEventManually()
     try {
-      val eventAsString = compact(render(JsonProtocolProxy.jsonProtocol.sparkEventToJson(event)))
-      streamingConnector.enqueueEvent(eventAsString, flush = flush, blocking = blocking)
+      streamingConnector.enqueueEvent(event, flush = flush, blocking = blocking)
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightListener.scala]: Why to repeat the names `flush` and `blocking`?

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 6]
```diff
@@ -127,7 +129,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
     * - Uses a dedicated httpClient to avoid collision with the one used by the main thread
     * - Payload is a Json Object containing the dm_app_id
     */
-  def sendHeartbeat(): Unit = {
+  def sendHeartbeat(): Unit = Utils.time(shouldLogDuration, "sendHeartbeat") {
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: The more I look at the change the often I ask myself this question: why not to use this `shouldLogDuration` inside `Utils.time` since that's the only reason for this `shouldLogDuration`. It's irrelevant to the class.

File: src/main/scala/co/datamechanics/delight/Utils.scala [hunk 7]
```diff
@@ -17,4 +19,20 @@ object Utils {
     bos.close()
     compressed
   }
+
+  def time[R](shouldLogDuration: Boolean, blockName: String)(block: => R): R = {
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/Utils.scala]: nit: it could be part of `delight` [package object](https://docs.scala-lang.org/tour/package-objects.html) which would make it available to all classes "implicitly".

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala (hunk 4 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala:267]: No worries! 
Actually the change is this PR just moves the costly operation of compacting the events in an async thread.

Before the change:
* Spark produces events and puts them in a queue with limited size
* Spark reads the queue and feeds it the Delight agent in a synchronous fashion
* The Delight agent compacts the events and puts them in its own queue
* The Delight agent queue is read by a separate thread which streams the events to the backend

After the change:
* Spark produces events and puts them in a queue with limited size
* Spark reads the queue and feeds it the Delight agent in a synchronous fashion
* The Delight agent puts the events in its own queue, without compacting them
* The Delight agent queue is read by a separate thread which compacts and streams the events to the backend

So basically, Spark does not need to wait for the agent to compact the events and so (hopefully) Spark queue will never reach its size limit.

What's the tradeoff? The Delight agent queue can know grow a lot (before it was the Spark queue). To be tested on super heavy job.

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala (hunk 6 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: You're right it's irrelevant to the class!
The reason why I did this is that `shouldLogDuration` is determined by the `sparkConf`, which is not accessible in the `Utils` package.

So an alternative would be to pass the `sparkConf` to `Utils.time` instead of passing it `shouldLogDuration`, but I'm not sure it's better?

File: src/main/scala/co/datamechanics/delight/Configs.scala (hunk 2 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/Configs.scala:50]: Do you mean the key should be in a Scala object as in the following?
```
val LOG_DURATION_KEY = "spark.delight.logDuration"
```

Comment (jaceklaskowski) [reply src/main/scala/co/datamechanics/delight/Configs.scala:50]: Yup. Or even like they did in [Spark](https://github.com/apache/spark/blob/084d38b64ecbcaa9fac47ffca5604cf2a72936fc/core/src/main/scala/org/apache/spark/internal/config/package.scala#L41-L52)

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala (hunk 6 above)
Comment (jaceklaskowski) [reply src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: I've got the same doubt. And I've got no other option, but as a code reader I think this change would make it slightly better.

File: src/main/scala/co/datamechanics/delight/Utils.scala (hunk 7 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/Utils.scala]: Fixed in [`8b0cc1b` (#9)](https://github.com/datamechanics/delight/pull/9/commits/8b0cc1b2ebf6c96ceb27adb2ccb05d9ec9723d42)

File: src/main/scala/co/datamechanics/delight/Configs.scala (hunk 2 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/Configs.scala:50]: I understand. Right now I'll avoid relying on the `ConfigBuilder` because it has changed between Spark 2.3 and Spark 3.1.
(I'm trying to limit the reliance on all Spark code: surprisingly we've been able to make a listener that works from Spark 2.3 to 3.1 thanks to stability of the SparkListener interface).

Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/Configs.scala:50]: I'm noting down this improvement for the future.

File: src/main/scala/co/datamechanics/delight/DelightListener.scala (hunk 5 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/DelightListener.scala]: Fixed in [`676cc41` (#9)](https://github.com/datamechanics/delight/pull/9/commits/676cc413d1684e7523eaa6c8095bf34b972a6cb9)

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala (hunk 1 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: Done in [`4936a15` (#9)](https://github.com/datamechanics/delight/pull/9/commits/4936a158611d5efd34f91744cd6e6383714b48a2)

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala (hunk 6 above)
Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: I'll postpone this one. If every call to `time` needs to extract the value of `shouldLogDuration` from the `sparkConf`, this will add some overhead to the calls to `time`.
A solution could be to create a `Timer` object that extracts the value from `sparkConf` at initialization time.
